cairo/crates/token/src/extensions/hyp_erc20_dex_collateral.cairo

```diff
@@ -0,0 +1,239 @@
+        #[flat]
+        RouterEvent: RouterComponent::Event,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        TokenRouterEvent: TokenRouterComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+        DexDeposit: DexDeposit,
+    }
+
+    // An event that is emitted when the DEX deposit is successful (via transfer_to_hook)
+    #[derive(Drop, starknet::Event)]
+    struct DexDeposit {
+        #[key]
+        token: ContractAddress,
+        #[key]
+        recipient: ContractAddress,
+        amount: u256,
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        mailbox: ContractAddress,
+        dex: ContractAddress,
+        wrapped_token: ContractAddress,
+        owner: ContractAddress,
+        hook: ContractAddress,
+        interchain_security_module: ContractAddress,
+    ) {
+        self.ownable.initializer(owner);
+        self
+            .mailbox
+            .initialize(mailbox, Option::Some(hook), Option::Some(interchain_security_module));
+        self.dex.write(dex);
+        self.collateral.initialize(wrapped_token);
+    }
+
+    impl TokenRouterHooksTraitImpl of TokenRouterHooksTrait<ContractState> {
+        fn transfer_from_sender_hook(
+            ref self: TokenRouterComponent::ComponentState<ContractState>, amount_or_id: u256,
+        ) -> Bytes {
+            HypErc20CollateralComponent::TokenRouterHooksImpl::transfer_from_sender_hook(
+                ref self, amount_or_id,
+            )
+        }
+
+        fn transfer_to_hook(
+            ref self: TokenRouterComponent::ComponentState<ContractState>,
+            recipient: u256,
+            amount_or_id: u256,
+            metadata: Bytes,
+        ) {
+            let recipient: ContractAddress = recipient.try_into().unwrap();
+            let amount: felt252 = amount_or_id.try_into().unwrap();
+
+            let mut contract_state = TokenRouterComponent::HasComponent::get_contract_mut(ref self);
+            let dex_address = contract_state.dex.read();
+            let token_address = contract_state.collateral.wrapped_token.read().contract_address;
+
+            // Approve the DEX to spend the tokens
+            let token_dispatcher = ERC20ABIDispatcher { contract_address: token_address };
+            token_dispatcher.approve(dex_address, amount_or_id);
+
+            let mut calldata = ArrayTrait::new();
+            recipient.serialize(ref calldata); // the actual recipient of the deposit
+            token_address.serialize(ref calldata); // depositing collateral token
+            amount.serialize(ref calldata);
+
+            let dex_call_result = call_contract_syscall(
+                address: dex_address,
+                entry_point_selector: DEX_DEPOSIT_ON_BEHALF_OF_SELECTOR,
+                calldata: calldata.span(),
+            );
+            assert(dex_call_result.is_ok(), 'DEPOSIT_FAILED');
```

Comment:
why cant we use a dispatcher similar to the ERC20 above?

Reply:
This is what Amay recommends https://hyperlaneworkspace.slack.com/archives/C08HMRXFR5Y/p1744302654335509?thread_ts=1744213764.557699&cid=C08HMRXFR5Y